### PR TITLE
feat(katana): added chunk size limit to events method

### LINKED ISSUE #2596: feat(katana): add chunk size limit for `starknet_getEvents` method
https://github.com/dojoengine/dojo/blob/8a9dc0cd874b65d4aeafd1b45b2108a44d954385/crates/katana/rpc/rpc/src/starknet/mod.rs#L494-L513

we should limit the chunk size to avoid getting dos by giving an insanely large value. the limit should be configurable through the cli

### PR DESCRIPTION
**Pull request by PoulavBhowmick03**
# Description

This PR adds a Chunk Size Limit to `starknet_getEvents` method

## Related issue

Fixes #2596 

## Tests

- [X] Yes
- [ ] No, because they aren't needed
- [ ] No, because I need help

## Added to documentation?

- [ ] README.md
- [ ] [Dojo Book](https://github.com/dojoengine/book)
- [X] No documentation needed

## Checklist

- [X] I've formatted my code (`scripts/prettier.sh`, `scripts/rust_fmt.sh`, `scripts/cairo_fmt.sh`)
- [ ] I've linted my code (`scripts/clippy.sh`, `scripts/docs.sh`)
- [ ] I've commented my code
- [ ] I've requested a review after addressing the comments

## Release Notes

- **New Features**
  - Introduced enhanced error handling for event page size with detailed error messages.
  - Added configuration options for maximum event page size in the RPC server.
  - Enhanced the `NodeArgs` struct to allow configuration of maximum event page size through command line arguments.
  - Implemented a default page size for RPC responses.
  - Added a new field for maximum event page size in the test configuration.
  
- **Bug Fixes**
  - Improved merging of command line arguments with configuration file settings for RPC options.

- **Documentation**
  - Updated comments and formatting for clarity throughout the codebase.

- **Tests**
  - Added new test cases to validate configuration merging behavior.

### CODE REVIEW

**Comment by PoulavBhowmick03** [comment]
@kariy Can you please take a look at this? Thanks!

File: crates/katana/rpc/rpc-types/src/error/starknet.rs [hunk 1]
```diff
@@ -83,6 +83,8 @@ pub enum StarknetApiError {
     TooManyKeysInFilter,
     #[error("Failed to fetch pending transactions")]
     FailedToFetchPendingTransactions,
+    #[error("Chunk Size is too big")]
+    ChunkSizeTooBig { requested: u64, maximum: u64 },
```
Comment (kariy) [inline crates/katana/rpc/rpc-types/src/error/starknet.rs:87]: we already have an error variant for this 

https://github.com/dojoengine/dojo/blob/0c70748fff67947895f3ff9255ef56ad986a5f41/crates/katana/rpc/rpc-types/src/error/starknet.rs#L34-L35

we can update the `PageSizeTooBig` variant into a struct variant and update the `.data()` function to include the struct fields,

https://github.com/dojoengine/dojo/blob/0c70748fff67947895f3ff9255ef56ad986a5f41/crates/katana/rpc/rpc-types/src/error/starknet.rs#L131-L145

File: crates/katana/rpc/rpc-types/src/error/starknet.rs [hunk 2]
```diff
@@ -118,6 +120,7 @@ impl StarknetApiError {
             StarknetApiError::UnsupportedContractClassVersion => 62,
             StarknetApiError::UnexpectedError { .. } => 63,
             StarknetApiError::ProofLimitExceeded => 10000,
+            StarknetApiError::ChunkSizeTooBig { .. } => 10001,
```
Comment (kariy) [inline crates/katana/rpc/rpc-types/src/error/starknet.rs:123]: remove as per earlier comment

File: crates/katana/rpc/rpc/src/starknet/mod.rs [hunk 3]
```diff
@@ -71,6 +71,7 @@ struct Inner<EF: ExecutorFactory> {
     block_producer: BlockProducer<EF>,
     blocking_task_pool: BlockingTaskPool,
     forked_client: Option<ForkedClient>,
+    max_chunk_size: u64,
```
Comment (kariy) [inline crates/katana/rpc/rpc/src/starknet/mod.rs:74]: lets make this field into an `Option` instead. also maybe lets create a config struct, `StarknetApiConfig`, to encapsulate this value.

Comment (kariy) [reply crates/katana/rpc/rpc/src/starknet/mod.rs:74]: we should also update all the `StarknetApi::new` functions to accept the new config struct

Comment (kariy) [reply crates/katana/rpc/rpc/src/starknet/mod.rs:74]: make sure to update the `RpcConfig` with the `max_events_page_size` option

https://github.com/dojoengine/dojo/blob/0c70748fff67947895f3ff9255ef56ad986a5f41/crates/katana/node/src/config/rpc.rs#L20-L28

i think 'page size' is a much better name in this case. maybe we should just call it page size anywhere else and just leave the chunk size name in the server impl.

File: crates/katana/rpc/rpc-types/src/error/starknet.rs (hunk 2 above)
Comment (glihm) [inline crates/katana/rpc/rpc-types/src/error/starknet.rs:123]: We might stick to the spec with `PageSizeTooBig`?
https://github.com/xJonathanLEI/starknet-rs/blob/ee9a5cdd678a3f1941599136432a1502ef772c00/starknet-core/src/types/codegen.rs#L1661

Or @kariy wdyt, are they semantically different?

**Review by kariy — COMMENTED** [review]
looks good so far. 

just need to make the value configurable from the cli.

**Review by kariy — COMMENTED** [review]
dont forget to add a test for this as well :)

File: bin/katana/src/cli/node.rs [hunk 4]
```diff
@@ -210,6 +210,8 @@ impl NodeArgs {
     }
 
     fn rpc_config(&self) -> RpcConfig {
+        const MAX_PAGE_SIZE: u64 = 10_000;
+        let page_size = Some(self.server.page_size.min(MAX_PAGE_SIZE));
```
Comment (kariy) [inline bin/katana/src/cli/node.rs:214]:
```suggestion
```

considering we already set a default value in the cli, i think its fine to not have this upper bound. i think we shouldn't impose constraints like this and let the power user do whatever they want with it. im also aware that leaving it unbounded can hurt the performance for user that is unaware of its implications. 

also if user set the value > 10k, this minimization will happen in silent, and thus may not be obvious for the user.

**Review by kariy — CHANGES_REQUESTED** [review]

**Comment by PoulavBhowmick03** [comment]
@kariy can you PTAL, it is not allowing me to resolve the merge conflicts for some reason, i have made the suggested changes

**Comment by glihm** [comment]
> @kariy can you PTAL, it is not allowing me to resolve the merge conflicts for some reason, i have made the suggested changes

You can always resolve merge conflicts by pulling the main branch from dojo and merging it into your branch locally. Then you push and your PR will be updated with merge conflicts revolved. 👍

File: crates/katana/node/src/lib.rs [hunk 5]
```diff
@@ -277,16 +278,20 @@ pub async fn spawn<EF: ExecutorFactory>(
     methods.register_method("health", |_, _| Ok(serde_json::json!({ "health": true })))?;
 
     if config.apis.contains(&ApiKind::Starknet) {
+        let starknet_api_config = StarknetApiConfig {
+            max_event_page_size: config.max_event_page_size.or(Some(1000)), // Default value
```
Comment (kariy) [inline crates/katana/node/src/lib.rs:282]:
```suggestion
            max_event_page_size: config.max_event_page_size
```

**Review by kariy — CHANGES_REQUESTED** [review]
all looks good except one small thing.

**Comment by kariy** [comment]
@PoulavBhowmick03 im happy to merge after the my comments and ci issues are addressed :)

**Comment by PoulavBhowmick03** [comment]
@kariy GM! Made the change.

**Comment by kariy** [comment]
pls run `cargo +nightly-2024-08-28 fmt --all` and `./scripts/clippy.sh`

**Comment by PoulavBhowmick03** [comment]
@kariy PTAL. done!

**Review by kariy — APPROVED** [review]

**Review by glihm — COMMENTED** [review]
Hey @PoulavBhowmick03, nice job on this!
Looks like a test is failing, you would mind trying to run it locally and see the issue if you can reproduce?

**Comment by glihm** [comment]
> @kariy @glihm there seems to be an error on the torii side? Or is it linked to my pr? Do let me know if it can be merged or do I need to make any changes!

Yeah, seems that torii used a value that is too big. Will have a look later in the day. 👍

**Comment by PoulavBhowmick03** [comment]
Any updates on this? :D
@kariy @glihm

**Comment by kariy** [comment]
> Any updates on this? :D @kariy @glihm

everything was working i just to bump the default value to match what torii was using. so decided to quickly did that :) thanks for the help on this

**Comment by PoulavBhowmick03** [comment]
> > Any updates on this? :D @kariy @glihm
> 
> everything was working i just had to bump the default value to match what torii was using in the tests (why it was failing). so decided to quickly did that :) thanks for the help on this

alright! thank you for letting me work on this!

**Pull request merged**